include/matx_tensor_ops.h

```diff
@@ -773,6 +773,171 @@ auto __MATX_INLINE__ remap(Op t, Ind idx, Inds... inds)
   return detail::RemapOp<DIM, decltype(op) , Ind>(op, idx);
 };   
 
+/**
+ * Slices elements from an operator/tensor.
+ */
+  namespace detail {
+  template <int DIM, typename T>
+  class SliceOp : public BaseOp<SliceOp<DIM, T>>
+  {
+    public: 
+      using scalar_type = typename T::scalar_type;
+      using shape_type = typename T::shape_type; 
+
+    private:
+      typename base_type<T>::type op_;
+      std::array<shape_type, DIM> sizes_;
+      std::array<shape_type, DIM> dims_;
+      std::array<shape_type, T::Rank()> starts_;
+      std::array<shape_type, T::Rank()> strides_;
+
+    public:
+      using matxop = bool;
+      using matxoplvalue = bool;
+
+      static_assert(T::Rank()>0, "SliceOp: Rank of operator must be greater than 0.");
+      static_assert(DIM<=T::Rank(), "SliceOp: DIM must be less than or equal to operator rank.");
+
+      __MATX_INLINE__ SliceOp(T op, const shape_type (&starts)[T::Rank()], const shape_type (&ends)[T::Rank()], const shape_type (&strides)[T::Rank()]) : op_(op) {
+        int d = 0;
+        for(int i = 0; i < T::Rank(); i++) {
+          shape_type start = starts[i];
+          shape_type end = ends[i];
+
+          starts_[i] = start;
+          strides_[i] = strides[i];
+
+          // compute dims and sizes
+          if(end != matxDropDim) {
+            dims_[d] = i;
+
+            if(end == matxEnd) {
+              sizes_[d] = op.Size(i) - start;
+            } else {
+              sizes_[d] = end - start;
+            }
+          
+	    //adjust size by stride
+            sizes_[d] = (shape_type)std::ceil(static_cast<double>(sizes_[d])/ static_cast<double>(strides_[d]));
+            d++;
+          }
+        }
+        MATX_ASSERT_STR(d==Rank(), matxInvalidDim, "SliceOp: Number of dimensions without matxDropDim must equal new rank.");
+      };
+
+      template <typename... Is>
+        __MATX_INLINE__ __MATX_DEVICE__ __MATX_HOST__ auto operator()(Is... indices) const 
+        {
+          static_assert(sizeof...(Is)==Rank());
```

Comment:
any chance we can have a version where there's no penalty for tensors?

Reply:
i'm not certain how that would look.  before we try to engineer around that we should first check if the overhead is significant or not.

Reply:
I'm not either, but every operator() access recomputing this would obviously be worse than what Slice does. not for this PR, but later it seems like we could figure it out.

Reply:
one option would be in the slice function to detect if op_ is a tensor and return op_.Slice(...); instead.  I'm still not convinced the overhead is sufficient for this extra path (x4) though.

Reply:
it's overhead that can be removed, plus it removes the need for Slice. it would be great if all those view functions were removed and still had no overhead for tensors.